tests/jsmarionette/runner/marionette-js-runner/host/session/device_session.js

```diff
@@ -0,0 +1,61 @@
+'use strict';
+let Device = require('./device');
+let Session = require('./session');
+let inherits = require('util').inherits;
+
+/**
+ * @constructor
+ */
+function DeviceSession(host, options) {
+  Session.apply(this, arguments);
+  this.device = new Device(host._deviceService.port);
+  this.profile = options.profile;
+}
+
+inherits(DeviceSession, Session);
+
+/**
+ * @return {Promise<void>}
+ *
+ * 1. Push profile to device.
+ * 2. Soft reboot device.
+ * 3. Start buffering logs.
+ */
+DeviceSession.prototype.start = function() {
+  let device = this.device;
+  return device.init()
+  .then(() => device.loadProfile(this.profile))
+  .then(() => device.rebootB2G())
+  .then(() => device.streamLogcat())
+  .then(logcat => logcat.pipe(this.host.log));
+};
+
+/**
+ * @return {Promise<void>}
+ */
+DeviceSession.prototype.destroy = function() {
+  this._logcat.removeAllListeners('data');
+  let device = this.device;
+  device._logStream.close();
+  return Promise.resolve();
+};
+
+/**
+ * @return {Promise<Error>}
+ */
+DeviceSession.prototype.checkError = function() {
+  let device = this.device;
```

Comment:
This replicates the issues we are seeing in bug 1175116 for both Devices and Desktop. I'm not sure how you want to go about making sure framework errors are reported but it would be nice to not get ourselves in the same problem. :)